README.md

```diff
@@ -291,173 +321,204 @@ be used in global variable initializers.
+
+```
+(string.measure_utf8 str:stringref pos:i32 codepoints:i32)
+  -> bytes:i32
+(string.measure_wtf8 str:stringref pos:i32 codepoints:i32)
   -> bytes:i32
 ```
 Measure how many bytes would be necessary to encode the contents of the
-string *`str`*, starting at cursor *`cursor`*, limited to a maximum of
-*`codepoints`* codepoints.  If the contents of the string cannot be
-encoded in the given encoding, return -1.
+string *`str`*, starting at position *`pos`*, to UTF-8 or WTF-8
+respectively, limited to *`codepoints`*.  For `string.measure_utf8`, if
+the *`codepoints`* following *`pos`* contain an isolated surrogate,
+return -1.
 
 The maximum number of bytes returned by `string.measure` is
 2<sup>31</sup>-1.  If an encoding would require more bytes, it is as if
 the contents can't be encoded at all (the return value is -1).
 
-(Note that the limit on cursor values constrains strings to have a
+(Note that the limit on position values constrains strings to have a
 maximum of 2<sup>31</sup>–1 codepoints, but it may be that such a string
 could exceed even 2<sup>32</sup> bytes in a given encoding; e.g. a
 string of 2<sup>31</sup>-1 `\uFFFF` codepoints would require
 2<sup>32</sup>+2<sup>31</sup>-3 bytes when encoded as UTF-8.)
 
+There is no `string.measure_wtf16`; `string.end_wtf16` provides the
+total code unit length, and any position slice in that range is valid
+and has a well-defined mapping to bytes.
+
 ```
-(string.encode $encoding $memory str:stringref cursor:i32 ptr:address bytes:i32)
-  -> cursor:i32, codepoints:i32, bytes:i32
+(string.encode_utf8 str:stringref pos:i32 bytes:i32)
+  -> bytes:i32
+(string.encode_wtf8 str:stringref pos:i32 bytes:i32)
+  -> bytes:i32
 ```
-Encode the contents of the string *`str`*, starting at the cursor
-*`cursor`*, to memory at *`ptr`*, limited to *`bytes`* bytes.  Return
-three values: a string cursor pointing to the remaining contents of the
-string, the number of codepoints written, and the number of bytes
+
+Encode the contents of the string *`str`* as UTF-8 or WTF-8,
+respectively, starting at the WTF-8 byte offset *`pos`*, to memory at
+*`ptr`*, limited to *`bytes`* bytes.  Return the number of bytes
 written.  Note that no `NUL` terminator is ever written.  If any
 codepoint cannot be encoded, trap.
 
 The maximum number of bytes that can be encoded at once by
 `string.encode` is 2<sup>31</sup>-1.  If an encoding would require more
 bytes, it is as if the codepoints can't be encoded (a trap).
 
+```
+(string.encode_wtf16 str:stringref pos:i32 codeunits:i32)
+```
+Encode the *`codeunits`* code units of the string *`str`* as WTF-16,
+respectively, starting at the WTF-16 code unit offset *`pos`*, to memory
+at *`ptr`*.  *`pos`*+*`codeunits`* must be a valid position in *`str`*.
+
 ### Predicates
 
 ```
-(string.eq a:stringref a-cursor:i32 b:stringref b-cursor:i32 codepoints:i32) -> i32
+(string.eq_wtf8 a:stringref a-pos:i32 b:stringref b-pos:i32 codepoints:i32) -> i32
```

Comment:
It might also be nice to have a `(string.eq a:stringref b:stringref)`, for what's likely the most common case of string comparisons: comparing entire strings. By skipping positions and length, this can be encoding-independent, which would avoid the issue that the encoding-specific variants will likely take a performance penalty on engines on which the respective encoding is foreign.

Reply:
This is equivalent to `(string.eq_wtf8 a 0 b 0 -1)` or `(string.eq_wtf16 a 0 b 0 -1)`, so it would just be an abbreviation and not imply transcoding.  Is that good enough?  Something is still not quite right here though, feels weird to have these two interfaces.

Reply:
I think what feels weird is that comparing two strings for equality conceptually shouldn't care about their encoding (because the same string could be encoded either way). But since the positions used as starting points have encoding-specific meaning, the current design needs the variants. Theoretically we'd even need at least one mixed variant, but I suppose we assume that wanting to mix wtf-8 and wtf-16 positions (`a: assumed-wtf8-stringref, a-pos:wtf8-codepoint-index, b: assumed-wtf16-string, b-pos: wtf16-codeunit-index`) won't happen in practice?